# (연습) 프로그래밍 맛보기

**주의 사항**

* 기존에 작성된 코드 셀과 텍스트 셀은 수정하지 않는다.
* 필요한 경우 코드 셀 또는 텍스트 셀을 추가해서 사용한다.
* 실습 파일을 제출할 때 구글 드라이브의 링크를 이용하여 누구나 공유할 수 있도록 설정한다.

**문제 1**

아래 명령문들을 실행해서 오류가 발생하는 경우 오류의 종류와 원인을 설명하고 수정하라.
오류가 발생하지 않으면 실행 결과를 설명한다.

(1)

In [1]:
print("Hello World)

SyntaxError: unterminated string literal (detected at line 1) (2771583741.py, line 1)

답

- 오류 종류: `SyntaxError` (구문 오류)
- 오류 원인: `print()` 함수의 인자로 사용된 문자열을 감싸는 큰따옴표가 시작 부분에만 사용되었음.
    따라서 닫는 큰따옴표를 추가해야 함.

In [4]:
print("Hello World")

Hello World


(2)

In [5]:
print(21 8)

SyntaxError: invalid syntax. Perhaps you forgot a comma? (1287684157.py, line 1)

답

- 오류 종류: `SyntaxError` (구문 오류)
- 오류 원인: `print()` 함수의 인자가 이상함. 218을 잘못 쓴 것으로 보임.

In [6]:
print(218)

218


(3)

In [7]:
print(2++2)

4


답

- 오류 발생하지 않음.
- `2 + 2`로 처림됨. 이유는 `+2`가 `2`로 처리되기 때문임.

**문제 2**

변수 할당 명령문과 관련하여 주의해야할 점들을 다룬다. 아래 명령문들을 실행해서 오류가 발생하는 경우 오류의 종류와 원인을 설명하고 수정한다.

(1)

In [8]:
x = 2;

답

- 오류 발생하지 않음.
- 세미 콜론 `;` 다음에 변수 할당문 등 간단한 명령문이 사용될 수 있지만 그냥 두어도 됨.

(2)

In [9]:
z = x y

SyntaxError: invalid syntax (1897392468.py, line 1)

답

- 오류 종류: `SyntaxError` (구문 오류)
- 오류 원인: `x y`는 의미가 없는 표현식임. 아마도 곱셈을 의도했을 수 있음.

In [10]:
y = 3
z = x * y
z

6

(3)

In [11]:
even = 3
odd = 17
print(even + oddd)

NameError: name 'oddd' is not defined

답

- 오류 종류: `NameError` (이름 오류).
- 오류 원인: 변수 `oddd`가 정의되어 있지 않음. 아마도 `odd`를 잘못 타이핑한 것으로 보임

In [12]:
even = 3
odd = 17
print(even + odd)

20


(4)

In [13]:
x = 3
zero = 0

print(x/zero)

ZeroDivisionError: division by zero

답

- 오류 종류: `ZeroDivisionError`
- 오류 원인: `zero` 변수가 0을 가리키기에 나눗셈의 분모가 0이되어 오류가 발생함.
    분모가 0이 되지 않도록 해야 함.

In [14]:
x = 3
zero = 0

print(x/(zero+2))

1.5


**문제 3**

언급된 값을 표현하는 수식을 표현한 다음에 실행하라.

(1) 정수 13에서 정수 19 빼기

힌트: 뺄셈 연산자를 활용한다.

답

In [15]:
13 - 19

-6

(2) 정수 81을 정수 3으로 나누기

힌트: 나눗셈 연산자를 활용한다.

답

In [16]:
81/3

27.0

(3) 정수 100을 정수 7로 나눈 나머지

힌트: 나머지 연산자를 활용한다.

답

In [17]:
100 % 7

2

(4) 정수 7.2의 세제곱

힌트: 거듭제곱 연산자를 활용한다.

답

In [18]:
7.2**3

373.24800000000005

소수점 이하 맨끝자리에서 보이는 5는 부동소수점 연산의 한계를 보여준다.
원래 0이어야 한다.

**문제 4**

먼저 아래와 같이 두 개의 변수를 선언한다.

In [19]:
width = 17
height = 12.0

언급된 값을 표현하는 수식을 두 변수를 이용하여 표현한 다음에 실행하여라.

(1) width를 2로 나눗 몫

답

In [20]:
width // 2

8

(2) width 나누기 2

답

In [21]:
width / 2

8.5

(3) width 나누기 2.0

답

In [22]:
width / 2.0

8.5

(4) height 나누기 3

답

In [23]:
height / 3

4.0

(5) 2 곱하기 5에 1 더하기

답

In [24]:
2 * 5 + 1

11

(6) 2 곱하기 5에 1.0 더하기

답

In [25]:
2 * 5 + 1.0

11.0

**문제 5**

오전 6시 52분에 집을 떠나서, 가볍게 (km당 8분 15초 소요) 1km를 뛰고, 
이어서 좀 빠르게 (km당 7분 12초 소요) 3km를 뛴 후, 다시 가볍게 1km를 마저 달려 집으로 돌아오면 
집에 도착하는 시간을 계산하여 아래처럼 출력하는 코드를 작성하라.

    집에 도착 시간: 7 시 30 분 6 초

힌트: 정수 나눗셈의 몫과 나머지 연산자 활용

**답**

시, 분, 초를 동시에 다뤄야 하기에 계산을 초 단위로 통일시킨다.
그렇지 않으면 시간의 연산이 복잡해진다.
먼저 출발시간인 6시 52분이 밤 0시를 기준으로 몇 초 지난 시간인지
계산해서 `start` 변수에 할당한다.

In [26]:
start = (6 * 60 + 52) * 60
start

24720

아래 두 변수는 1 킬로미터를 가볍게 달릴 때와 좀 빠르게 달릴 때 걸리는 시간(초)을 가리킨다.

In [27]:
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

그러면 달리는 시간을 초로 계산하면 2286초로 나온다.

In [28]:
running_time = (1 + 1) * slow_running + 3 * fast_running
running_time

2286

도착시간은 밤 0시를 기준으로 출발시간에 달린 시간을 더하면 된다.

In [29]:
arrival = start + running_time
arrival

27006

도착시간을 시, 분, 초로 변환하기 위해 1분은 60초, 1시간은 60분이라는 정보를
이용하여 도착시간에 대해 60으로 나눴을 때의 몫과 나머지 계산을 두 번 연속 적용한다.

- 도착시간의 초 계산: 60으로 나눴을 때의 나머지

In [30]:
arrival_seconds = arrival % 60
arrival_seconds

6

- 도착시간의 분 계산: 60으로 나눴을 때의 몫

In [31]:
arrival_minutes = arrival // 60
arrival_minutes

450

450분은 60분 보다 크기에 시와 분으로 다시 변환해야 한다.
한 시간이 60분이기에 역시 60으로 나눈 몫과 나머지를 계산한다.

- 도착시간의 시 계산: 60으로 나눴을 때의 몫

In [32]:
arrival_hours = arrival_minutes // 60
arrival_hours

7

- 도착시간의 분 계산: 60으로 나눴을 때의 나머지

In [33]:
arrival_minutes = arrival_minutes % 60
arrival_minutes

30

따라서 도착시간은 7시 30분 6초이다. 

In [34]:
print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초


정리하면 다음과 같다.

In [35]:
start = (6 * 60 + 52) * 60
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

running_time = (1 + 1) * slow_running + 3 * fast_running
arrival = start + running_time

arrival_seconds = arrival % 60
arrival_minutes = arrival // 60
arrival_hours = arrival_minutes // 60
arrival_minutes = arrival_minutes % 60

print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초
